Mathlib/Topology/UniformSpace/AbstractCompletion.lean

```diff
@@ -0,0 +1,405 @@
+/-
+Copyright (c) 2019 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot
+
+! This file was ported from Lean 3 source module topology.uniform_space.abstract_completion
+! leanprover-community/mathlib commit dc6c365e751e34d100e80fe6e314c3c3e0fd2988
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Topology.UniformSpace.UniformEmbedding
+import Mathlib.Topology.UniformSpace.Equiv
+
+/-!
+# Abstract theory of Hausdorff completions of uniform spaces
+
+This file characterizes Hausdorff completions of a uniform space α as complete Hausdorff spaces
+equipped with a map from α which has dense image and induce the original uniform structure on α.
+Assuming these properties we "extend" uniformly continuous maps from α to complete Hausdorff spaces
+to the completions of α. This is the universal property expected from a completion.
+It is then used to extend uniformly continuous maps from α to α' to maps between
+completions of α and α'.
+
+This file does not construct any such completion, it only study consequences of their existence.
+The first advantage is that formal properties are clearly highlighted without interference from
+construction details. The second advantage is that this framework can then be used to compare
+different completion constructions. See `topology/uniform_space/compare_reals` for an example.
+Of course the comparison comes from the universal property as usual.
+
+A general explicit construction of completions is done in `uniform_space/completion`, leading
+to a functor from uniform spaces to complete Hausdorff uniform spaces that is left adjoint to the
+inclusion, see `uniform_space/UniformSpace` for the category packaging.
+
+## Implementation notes
+
+A tiny technical advantage of using a characteristic predicate such as the properties listed in
+`abstract_completion` instead of stating the universal property is that the universal property
+derived from the predicate is more universe polymorphic.
```

Comment:
```suggestion
`AbstractCompletion` instead of stating the universal property is that the universal property
derived from the predicate is more universe polymorphic.
```